File: backend/models/Transaction.js
const mongoose = require("mongoose");

const transactionSchema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
    },
    type: {
      type: String,
      enum: ["deposit", "withdrawal", "payment", "refund", "fee", "commission"],
      required: true,
    },
    amount: {
      type: Number,
      required: true,
    },
    currency: {
      type: String,
      enum: ["USD", "SYP"],
      required: true,
    },
    status: {
      type: String,
      enum: ["pending", "completed", "failed", "cancelled"],
      default: "pending",
    },
    method: {
      type: String,
      enum: ["wallet", "cash", "card", "bank-transfer", "mobile-payment"],
      required: true,
    },
    reference: {
      type: String,
      unique: true,
      sparse: true,
    },
    relatedShipment: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Shipment",
    },
    description: {
      type: String,
      default: "",
    },
    balanceBefore: {
      type: Number,
      required: true,
    },
    balanceAfter: {
      type: Number,
      required: true,
    },
    metadata: {
      type: Map,
      of: String,
    },
    processedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
    },
    processedAt: Date,
  },
  {
    timestamps: true,
  }
);

// Generate reference number
transactionSchema.pre("save", async function (next) {
  if (!this.reference) {
    const prefix = "TXN";
    const random = Math.random().toString(36).substring(2, 10).toUpperCase();
    const timestamp = Date.now().toString().slice(-8);
    this.reference = `${prefix}${timestamp}${random}`;
  }
  next();
});

module.exports = mongoose.model("Transaction", transactionSchema);

File: flutter_webview/lib/models/transaction.dart
import 'shipment.dart';

class WalletTransaction {
  final String id;
  final String type;
  final String status;
  final String method;
  final double amount;
  final String currency;
  final String description;
  final DateTime createdAt;
  final Shipment? relatedShipment;

  WalletTransaction({
    required this.id,
    required this.type,
    required this.status,
    required this.method,
    required this.amount,
    required this.currency,
    required this.description,
    required this.createdAt,
    this.relatedShipment,
  });

  factory WalletTransaction.fromJson(Map<String, dynamic> json) {
    final relatedShipmentJson = json['relatedShipment'];
    return WalletTransaction(
      id: json['_id']?.toString() ?? json['id']?.toString() ?? '',
      type: json['type']?.toString() ?? '',
      status: json['status']?.toString() ?? '',
      method: json['method']?.toString() ?? '',
      amount: (json['amount'] is num)
          ? (json['amount'] as num).toDouble()
          : double.tryParse(json['amount']?.toString() ?? '0') ?? 0,
      currency: json['currency']?.toString() ?? '',
      description: json['description']?.toString() ?? '',
      createdAt: DateTime.tryParse(json['createdAt']?.toString() ?? '') ??
          DateTime.now(),
      relatedShipment: relatedShipmentJson is Map<String, dynamic>
          ? Shipment.fromJson(relatedShipmentJson)
          : null,
    );
  }
}
